client/src/main/java/org/spine3/validate/ValidatingBuilder.java

```diff
@@ -40,16 +52,117 @@
      * @param fieldValue the value of the field
      * @param fieldName  the name of the field
      * @param <V>        the type of the field value
-     * @throws ConstraintViolationThrowable if there are some constraint violations
+     * @throws ConstraintViolationThrowable if there are any constraint violations
      */
     <V> void validate(FieldDescriptor descriptor,
                       V fieldValue,
                       String fieldName) throws ConstraintViolationThrowable;
 
     /**
-     * Returns constructed and validated {@code Message}.
+     * Validates and builds {@code Message}.
      *
      * @return the {@code Message} instance
+     * @throws ConstraintViolationThrowable if there are any constraint violations
+     */
+    T build() throws ConstraintViolationThrowable;
+
+    /**
+     * Merges the message currently built with the fields of a given {@code message}.
+     *
+     * @param message the message to merge into the state this builder
+     * @return the builder instance
+     */
+    ValidatingBuilder<T, B> mergeFrom(T message);
+
+    /**
+     * Sets an original state for this builder to allow building the new value on top of some
+     * pre-defined value.
+     *
+     * <p>The validation of the passed {@code state} is NOT performed.
+     *
+     * @param state the new state
+     */
+    @Internal
+    void setOriginalState(T state);
+
+    /**
+     * Determines if the current message state has been modified comparing to its original state.
+     *
+     * @return {@code true} if it is modified, {@code false} otherwise
+     */
+    @Internal
+    boolean isDirty();
+
+    /**
+     * Clears the state of this builder.
+     *
+     * <p>In particular, the state of the message being built is reset to the default.
+     * Also the original state (if it {@linkplain #setOriginalState(Message) has been set}
+     * previously) is cleared.
      */
-    T build();
+    @Internal
+    void clear();
+
+    /**
+     * Enumeration of generic type parameters of this interface.
+     */
+    enum GenericParameter {
```

Comment:
Please move `GenetricTypeIndex` to the client side. This is already one case when we need it, and there may be other as we develop UI-related technology.